File: analysis/R/03_networks.R
# ---------------------------------------------------------------------------
# 03 - Networks: co-membership degree and regional assortativity.
#
#   Rscript analysis/R/03_networks.R
#
# Base R only. igraph is not required; if it happens to be installed the script
# uses it for community detection, and says so when it is not. Everything up to
# that point runs on a bare R installation.
# ---------------------------------------------------------------------------

govtn_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg)) {
    return(dirname(normalizePath(sub("^--file=", "", file_arg[1]))))
  }
  frame <- sys.frame(1)$ofile
  if (!is.null(frame)) return(dirname(normalizePath(frame)))
  getwd()
}
source(file.path(govtn_script_dir(), "load_govtn.R"))

out_dir <- file.path(dirname(dirname(govtn_script_dir())), "output", "tables")
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

persons <- govtn_load("persons")
edges   <- govtn_load("edges_co_membership")

cat(sprintf("Co-membership network: %d nodes, %d edges\n",
            nrow(persons), nrow(edges)))
cat("Edge weight is DAYS OF OVERLAPPING SERVICE, not a count of shared cabinets.\n")

# --- Degree and weighted degree --------------------------------------------
# The edge list is undirected and stored once per pair, so each endpoint has to
# be counted on both sides.
endpoints <- c(edges$source, edges$target)
weights   <- c(edges$weight, edges$weight)

degree <- table(endpoints)
strength <- tapply(weights, endpoints, sum)

central <- data.frame(
  person_id = names(degree),
  degree = as.integer(degree),
  weighted_degree = as.numeric(strength[names(degree)]),
  stringsAsFactors = FALSE
)
central <- merge(central, persons[, c("person_id", "name", "birth_governorate",
                                      "n_appointments")],
                 by = "person_id", all.x = TRUE)
central <- central[order(-central$degree), ]

cat("\nMost connected ministers by co-membership degree\n")
print(head(central[, c("name", "degree", "weighted_degree", "n_appointments")], 10),
      row.names = FALSE)
write.csv(central, file.path(out_dir, "03_centrality.csv"),
          row.names = FALSE, fileEncoding = "UTF-8")

# --- Regional assortativity -------------------------------------------------
# Nominal assortativity on region of birth: a direct test of regional closure.
#
# Use birth_governorate, NOT birth_region. The latter is Wikidata's raw P131
# label, which names a delegation - several hundred near-unique categories that
# drive the coefficient towards zero and make closure look absent.
#
# Vertices with no coded governorate are dropped rather than pooled: ministers
# born abroad have no governorate by design, and letting NA become its own
# category would count "born abroad" as a region and inflate the coefficient.
attr_of <- setNames(persons$birth_governorate, persons$person_id)
a <- attr_of[edges$source]
b <- attr_of[edges$target]
keep <- !is.na(a) & !is.na(b)
a <- a[keep]; b <- b[keep]
w <- edges$weight[keep]

cat(sprintf("\nEdges with both endpoints coded: %d of %d\n", sum(keep), nrow(edges)))

# Newman's nominal assortativity, computed on the edge mixing matrix.
levels_all <- sort(unique(c(a, b)))
e <- matrix(0, length(levels_all), length(levels_all),
            dimnames = list(levels_all, levels_all))
for (i in seq_along(a)) {
  # Each undirected edge contributes to both cells, so the matrix is symmetric.
  e[a[i], b[i]] <- e[a[i], b[i]] + w[i]
  e[b[i], a[i]] <- e[b[i], a[i]] + w[i]
}
e <- e / sum(e)
tr <- sum(diag(e))
sq <- sum(rowSums(e) * colSums(e))
r_weighted <- (tr - sq) / (1 - sq)

# Unweighted, so the result does not depend on tenure length.
e2 <- matrix(0, length(levels_all), length(levels_all),
             dimnames = list(levels_all, levels_all))
for (i in seq_along(a)) {
  e2[a[i], b[i]] <- e2[a[i], b[i]] + 1
  e2[b[i], a[i]] <- e2[b[i], a[i]] + 1
}
e2 <- e2 / sum(e2)
r_plain <- (sum(diag(e2)) - sum(rowSums(e2) * colSums(e2))) /
           (1 - sum(rowSums(e2) * colSums(e2)))

cat(sprintf("Assortativity by governorate of birth: %.4f (unweighted), %.4f (by overlap days)\n",
            r_plain, r_weighted))
cat("Near zero means ministers from the same governorate are no more likely to\n")
cat("serve together than chance - co-membership is set by cabinet timing, not origin.\n")

write.csv(
  data.frame(measure = c("assortativity_unweighted", "assortativity_weighted"),
             value = c(r_plain, r_weighted)),
  file.path(out_dir, "03_assortativity.csv"),
  row.names = FALSE, fileEncoding = "UTF-8")

# --- Optional: communities --------------------------------------------------
if (requireNamespace("igraph", quietly = TRUE)) {
  g <- igraph::graph_from_data_frame(
    edges[, c("source", "target", "weight")], directed = FALSE)
  communities <- igraph::cluster_louvain(g, weights = igraph::E(g)$weight)
  cat(sprintf("\nLouvain modularity: %.3f across %d communities\n",
              igraph::modularity(communities), length(communities)))
} else {
  cat("\n(igraph not installed - skipping community detection.",
      "install.packages(\"igraph\") to enable it.)\n")
}

cat("\nWrote 2 tables to ", out_dir, "\n", sep = "")

File: analysis/python/03_networks.py
"""03 - Networks: co-membership degree and regional assortativity.

    python analysis/python/03_networks.py

Needs pandas. networkx is optional; if installed the script adds community
detection, and says so when it is not. `analysis/R/03_networks.R` is the same
computation in base R and produces the same coefficients.
"""

import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))

import pandas as pd  # noqa: E402

from load_govtn import load  # noqa: E402

REPO = pathlib.Path(__file__).resolve().parents[2]
OUT = REPO / "output" / "tables"
OUT.mkdir(parents=True, exist_ok=True)

persons = load("persons")
edges = load("edges_co_membership")

print(f"Co-membership network: {len(persons)} nodes, {len(edges)} edges")
print("Edge weight is DAYS OF OVERLAPPING SERVICE, not a count of shared cabinets.")

# --- Degree and weighted degree --------------------------------------------
# The edge list is undirected and stored once per pair, so each endpoint has to
# be counted on both sides.
endpoints = pd.concat([edges.source, edges.target], ignore_index=True)
weights = pd.concat([edges.weight, edges.weight], ignore_index=True)

central = (pd.DataFrame({"person_id": endpoints, "weight": weights})
           .groupby("person_id")
           .agg(degree=("weight", "size"), weighted_degree=("weight", "sum"))
           .reset_index()
           .merge(persons[["person_id", "name", "birth_governorate",
                           "n_appointments"]], on="person_id", how="left")
           .sort_values("degree", ascending=False))

print("\nMost connected ministers by co-membership degree")
print(central[["name", "degree", "weighted_degree", "n_appointments"]]
      .head(10).to_string(index=False))
central.to_csv(OUT / "03_centrality.csv", index=False)

# --- Regional assortativity -------------------------------------------------
# Nominal assortativity on region of birth: a direct test of regional closure.
#
# Use birth_governorate, NOT birth_region. The latter is Wikidata's raw P131
# label, which names a delegation - several hundred near-unique categories that
# drive the coefficient towards zero and make closure look absent.
#
# Vertices with no coded governorate are dropped rather than pooled: ministers
# born abroad have no governorate by design, and letting NA become its own
# category would count "born abroad" as a region and inflate the coefficient.
attr_of = dict(zip(persons.person_id, persons.birth_governorate))
a = edges.source.map(attr_of)
b = edges.target.map(attr_of)
keep = a.notna() & b.notna()
print(f"\nEdges with both endpoints coded: {int(keep.sum())} of {len(edges)}")


def assortativity(left: pd.Series, right: pd.Series,
                  weight: pd.Series | None = None) -> float:
    """Newman's nominal assortativity from the edge mixing matrix."""
    w = pd.Series(1.0, index=left.index) if weight is None else weight.astype(float)
    frame = pd.DataFrame({"a": left, "b": right, "w": w})
    # Each undirected edge contributes to both cells, so the matrix is symmetric.
    both = pd.concat([frame, frame.rename(columns={"a": "b", "b": "a"})])
    mixing = both.pivot_table(index="a", columns="b", values="w",
                              aggfunc="sum", fill_value=0.0)
    mixing = mixing.reindex(index=mixing.columns.union(mixing.index),
                            columns=mixing.columns.union(mixing.index),
                            fill_value=0.0)
    e = mixing.to_numpy()
    e = e / e.sum()
    trace = e.trace()
    squared = float((e.sum(axis=1) * e.sum(axis=0)).sum())
    return (trace - squared) / (1 - squared)


r_plain = assortativity(a[keep], b[keep])
r_weighted = assortativity(a[keep], b[keep], edges.weight[keep])
print(f"Assortativity by governorate of birth: {r_plain:.4f} (unweighted), "
      f"{r_weighted:.4f} (by overlap days)")
print("Near zero means ministers from the same governorate are no more likely to")
print("serve together than chance - co-membership is set by cabinet timing, not origin.")

pd.DataFrame({"measure": ["assortativity_unweighted", "assortativity_weighted"],
              "value": [r_plain, r_weighted]}).to_csv(
    OUT / "03_assortativity.csv", index=False)

# --- Optional: communities --------------------------------------------------
try:
    import networkx as nx

    graph = nx.from_pandas_edgelist(edges, "source", "target", ["weight"])
    communities = nx.community.louvain_communities(graph, weight="weight", seed=1)
    modularity = nx.community.modularity(graph, communities, weight="weight")
    print(f"\nLouvain modularity: {modularity:.3f} across {len(communities)} communities")
except ImportError:
    print("\n(networkx not installed - skipping community detection. "
          "pip install networkx to enable it.)")

print(f"\nWrote 2 tables to {OUT}")
